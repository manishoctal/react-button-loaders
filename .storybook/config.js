import { configure } from '@storybook/react';
import { injectGlobal } from 'styled-components';

injectGlobal`
  @import url('https://fonts.googleapis.com/css?family=Heebo:400,600');
  button span {
    font-family: 'Heebo', sans-serif;
    font-weight: 600;
  }
`
// automatically import all files ending in *.stories.js
const req = require.context('../src/stories', true, /.stories.js$/);
function loadStories() {
  req.keys().forEach((filename) => req(filename));
}

configure(loadStories, module);
