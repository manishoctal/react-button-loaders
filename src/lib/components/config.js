export default {
  prefixCls: 'react-button-loaders'
}
