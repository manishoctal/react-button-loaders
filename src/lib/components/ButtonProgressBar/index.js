import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import config from '../../config.js'
const { prefixCls } = config
import {
  ButtonStyled,
  Text,
  Loading,
  Success,
  RefreshIcon,
  CheckIcon
} from './style'

const ButtonProgressBar = ({
  children,
  className,
  state,
  refreshComponent,
  checkComponent,
  textWhileLoading,
  textWhenLoaded,
  ...props
}) => {
  const classNameJoined = classNames(className, `${prefixCls}__btn`, {
    [`${prefixCls}__loading`]: state === 'loading',
    [`${prefixCls}__finished`]: state === 'finished'
  })
  return (
    <ButtonStyled {...props} className={classNameJoined}>
      <Text>{children}</Text>
      <Loading>
        {refreshComponent}
        <span>{textWhileLoading}</span>
      </Loading>
      <Success>
        {checkComponent}
        <span>{textWhenLoaded}</span>
      </Success>
    </ButtonStyled>
  )
}

ButtonProgressBar.defaultProps = {
  className: '',
  speedProgress: 3000,
  speedIconLoader: 800,
  refreshComponent: <RefreshIcon />,
  checkComponent: <CheckIcon />,
  onClick: () => {},
  bgColor: '#fd5c63',
  bgLoading: '#3594ca',
  bgLoadingBehind: '#3785b0'
}

const { string, number, func } = PropTypes

ButtonProgressBar.propTypes = {
  textWhenLoaded: string,
  textWhileLoading: string,
  className: string,
  speedProgress: number,
  speedIconLoader: number,
  onClick: func,
  bgColor: string,
  bgLoading: string,
  bgLoadingBehind: string
}

export default ButtonProgressBar
