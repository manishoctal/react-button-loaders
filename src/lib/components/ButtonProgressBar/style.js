import React from 'react'
import styled, { keyframes } from 'styled-components'
import { IconRefresh, IconCheck } from '../Icons'
import config from '../../config'
const { prefixCls } = config

const rotate360 = keyframes`
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
`

const progressBar = keyframes`
  from { transform: translateX(-100%) }
  to { transform: translateX(0) }
`

const scaleFinished = keyframes`
	0%   { transform: scale(10) }
	50%  { transform: scale(0.2) }
	70%  { transform: scale(1.2) }
	90%  { transform: scale(0.7) }
	100% { transform: scale(1) }
`

export const Text = styled.span`
  font-size: 15px;
  font-weight: 300;
`
Text.displayName = 'Text'

export const Loading = styled.span``
Loading.displayName = 'Loading'

export const Success = styled.span``
Success.displayName = 'Success'

const RefreshStyled = styled.span`
  svg {
    fill: #fff;
    width: 20px;
  }
`

const CheckStyled = styled.span`
  svg {
    stroke: #FFF
    width: 24px;
    height: 24px;
  }
`

export const RefreshIcon = () => (
  <RefreshStyled>
    <IconRefresh />
  </RefreshStyled>
)

export const CheckIcon = () => (
  <CheckStyled>
    <IconCheck />
  </CheckStyled>
)

export const ButtonStyled = styled.button`
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  width: 200px;
  padding: 24px 50px;
  background-color: #03a9f4;
  color: #fff;
  border: none;
  border-radius: 4px;
  box-shadow: 0 12px 24px 0 rgba(0, 0, 0, 0.2);
  text-transform: uppercase;
  overflow: hidden;
  outline: none;
  &:before {
    position: absolute;
    content: '';
    bottom: 0;
    left: 0;
    transform: translateX(-100%);
    width: 100%;
    height: 100%;
    background-color: #54d98c;
  }
  ${Text}, ${Loading}, ${Success} {
    position: absolute;
    line-height: 0;
  }
  ${Text} {
    top: 50%;
    transform: translateY(-50%);
  }
  ${Loading} {
    top: 100%;
    transform: translateY(0%);
  }
  ${Success} {
    display: none;
  }

  &.${prefixCls}__loading {
    cursor: progress;
    pointer-events: none;
    background-color: #2ecc71;
    &:before {
      animation: ${progressBar} ${props => props.speedProgress}ms linear
        infinite;
    }

    span {
      &:nth-of-type(1) {
        top: -100%;
        transform: translateY(-50%);
      }

      &:nth-of-type(2) {
        top: 50%;
        transform: translateY(-50%);

        svg {
          animation: ${rotate360} ${props => props.speedIconLoader}ms linear
            infinite;
        }
      }

      &:nth-of-type(3) {
        display: none;
      }
    }
  }
  &.${prefixCls}__finished {
    background-color: #54d98c;
    pointer-events: none;
    ${Text}, ${Loading} {
      display: none;
    }

    ${Success} {
      display: block !important;
      animation: ${scaleFinished} 0.5s linear;
    }
  }
`
