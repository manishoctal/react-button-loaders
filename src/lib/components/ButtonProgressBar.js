import React from 'react'
import classNames from 'classnames'
import config from './config.js'
const { prefixCls } = config
import { ButtonStyled } from './style'
import { IconRefresh, IconCheck } from './Icons'

const ButtonProgressBar = ({
  children,
  className,
  loading,
  finished,
  onClick,
  onFinished
}) => {
  const classNameJoined = classNames(className, `${prefixCls}__btn`, {
    [`${prefixCls}__loading`]: loading,
    [`${prefixCls}__finished`]: finished
  })

  return (
    <ButtonStyled
      {...this.props}
      className={classNameJoined}
      onClick={onClick}
      onTransitionEnd={onFinished}
    >
      <span className={`${prefixCls}__text`}>{children}</span>
      <span className={`${prefixCls}__loading`}>
        <IconRefresh />
      </span>
      <span className={`${prefixCls}__check`}>
        <IconCheck />
      </span>
    </ButtonStyled>
  )
}

ButtonProgressBar.defaultProps = {
  className: '',
  onClick: () => {},
  onFinished: () => {}
}

export default ButtonProgressBar
