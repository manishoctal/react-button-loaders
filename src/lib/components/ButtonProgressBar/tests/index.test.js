import React from 'react'
import config from '../../../config.js'
const { prefixCls } = config
import ButtonProgressBar from '../index'
import { shallow } from 'enzyme'

describe('ButtonProgressBar component', () => {
  test('should render a the button progress bar', () => {
    const wrapper = shallow(<ButtonProgressBar />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should call mock function when button is clicked', () => {
    const handleClick = jest.fn()
    const wrapper = shallow(<ButtonProgressBar onClick={handleClick} />)
    wrapper.simulate('click')
    expect(handleClick).toHaveBeenCalled()
  })

  test('should put class `__loading` when the state is loading ', () => {
    const wrapper = shallow(<ButtonProgressBar state='loading' />)
    expect(wrapper.find(`.${prefixCls}__loading`).length).toEqual(1)
  })

  test('should put class `__finished` when the state is finished ', () => {
    const wrapper = shallow(<ButtonProgressBar state='finished' />)
    expect(wrapper.find(`.${prefixCls}__finished`).length).toEqual(1)
  })
})
