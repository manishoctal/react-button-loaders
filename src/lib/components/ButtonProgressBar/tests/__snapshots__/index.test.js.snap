// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonProgressBar component should render a the button progress bar 1`] = `
<styled.button
  bgColor="#fd5c63"
  bgLoading="#3594ca"
  bgLoadingBehind="#3785b0"
  className="react-button-loaders__btn"
  onClick={[Function]}
  speedIconLoader={800}
  speedProgress={3000}
>
  <Text />
  <Loading>
    <Component />
  </Loading>
  <Success>
    <Component />
  </Success>
</styled.button>
`;
